### ISSUE TITLE
dapinput.ini unable to assign SHIFT, CTRL, ALT, ARROW KEYS

### ISSUE THREAD

**Issue by OzFalcon** [btolab/hypseus#16]
When using dapinput.ini file to assign keys to switches, 
The following keys are unable to be assigned to any switches
> Left Alt, Right Alt
> Left Ctrl, Right Ctrl
> Left Shift, Right Shift
> Up, Down, Left, Right Arrow keys.

Attached is a typical full key layout dapinput.ini where these keys are required.
[dapinput.ini.full.txt](https://github.com/btolab/hypseus/files/1837366/dapinput.ini.full.txt)

**Comment by ghost** [comment]
the  input system does need an update for the  the analogue joystick and joystick buttons. The controls are moot at the moment for joysticks.  I didnt notice that particular problems with the keyboard but ill look into it please see here for the joystick fixes. I done for daphne 

https://github.com/RetroPie/daphne-emu/pull/15

**Comment by OzFalcon** [comment]
Forgot to mention,
If these keys are attempted to be assigned, Then the switch ends up with no assignment at all.

**Comment by ghost** [comment]
The keys should assign but for them to work you need to override the defaults. If you want to use the arrow keys set up down left and right to something else.  Defaults are as follows

int g_key_defs[SWITCH_COUNT][2] = {
    {SDLK_UP, SDLK_KP_8},     // up
    {SDLK_LEFT, SDLK_KP_4},   // left
    {SDLK_DOWN, SDLK_KP_2},   // down
    {SDLK_RIGHT, SDLK_KP_6},  // right
    {SDLK_1, 0},                      // 1 player start
    {SDLK_2, 0},                      // 2 player start
    {SDLK_SPACE, SDLK_LCTRL}, // action button 1
    {SDLK_LALT, 0},                   // action button 2
    {SDLK_LSHIFT, 0},                 // action button 3
    {SDLK_5, SDLK_c},         // coin chute left
    {SDLK_6, 0},                      // coin chute right
    {SDLK_KP_DIVIDE, 0},              // skill easy
    {SDLK_KP_MULTIPLY, 0},            // skill medium
    {SDLK_KP_MINUS, 0},               // skill hard
    {SDLK_9, 0},                      // service coin
    {SDLK_F2, 0},                     // test mode
    {SDLK_F3, 0},                     // reset cpu
    {SDLK_F12, SDLK_F11},     // take screenshot
    {SDLK_ESCAPE, SDLK_q},    // Quit DAPHNE
    {SDLK_p, 0},                      // pause game
    {SDLK_BACKQUOTE, 0},                  // toggle console (TODO)
    {SDLK_t, 0},                      // Tilt/Slam switch
};

**Comment by OzFalcon** [comment]
No, They don't assign. It's specifically those keys listed (CTRL, ALT, SHIFT, ARROWS)
The other keys assign fine.

Try this simple example test and you will see pause SHOULD be assigned to up arrow.
Even though you can assign a key to multiple switches - In this example we explicitly redefine
"KEY_UP" to the "i" key to ensure there are no conflicts. Running shows pause is not bound to "Up arrow".
```
[KEYBOARD]
KEY_UP = 105 0 0
KEY_PAUSE = 273 0 0
END
```

**Comment by ghost** [comment]
ill look into that the joystick had problems with buttons doing that. Ill look at the code when i get off work should be easy enough to fix. 

nt g_key_defs[SWITCH_COUNT][2] = {
{SDLK_UP, SDLK_KP_8}, // up
{SDLK_LEFT, SDLK_KP_4}, // left
{SDLK_DOWN, SDLK_KP_2}, // down
{SDLK_RIGHT, SDLK_KP_6}, // right
{SDLK_1, 0}, // 1 player start
{SDLK_2, 0}, // 2 player start
{SDLK_SPACE, SDLK_LCTRL}, // action button 1
{SDLK_LALT, 0}, // action button 2
{SDLK_LSHIFT, 0}, // action button 3
{SDLK_5, SDLK_c}, // coin chute left
{SDLK_6, 0}, // coin chute right
{SDLK_KP_DIVIDE, 0}, // skill easy
{SDLK_KP_MULTIPLY, 0}, // skill medium
{SDLK_KP_MINUS, 0}, // skill hard
{SDLK_9, 0}, // service coin
{SDLK_F2, 0}, // test mode
{SDLK_F3, 0}, // reset cpu
{SDLK_F12, SDLK_F11}, // take screenshot
{SDLK_ESCAPE, SDLK_q}, // Quit DAPHNE
{SDLK_p, 0}, // pause game
{SDLK_BACKQUOTE, 0}, // toggle console (TODO)
{SDLK_t, 0}, // Tilt/Slam switch
};

**Comment by ghost** [comment]
try this 
[KEYBOARD]
KEY_UP = 105 0 0
KEY_PAUSE = 82 0 0
END

https://gist.github.com/stuartpb/804078/eb8df5f6dbfd063401b5a67a6a2509ed64db5b66

if this doesnt work ill check whats going on in the keyboard code in sdl2

**Comment by OzFalcon** [comment]
ASCII Code 82 aka Capital "R" Does not work (Same as SHIFT, CTRL, ALT, ARROWS).
I'd take a guess that all the Capitals don't work.

**Comment by ghost** [comment]
try up with code 82 for up forget ascii for now.  

see 

https://github.com/davidsiaw/SDL2/blob/master/include/SDL_keycode.h

looks like its just expanding a macro for keys other than standards ascii. ill look into thsi properly when i get the time wont be too hard to fix

**Comment by OzFalcon** [comment]
Ok, got it.
https://wiki.libsdl.org/SDLKeycodeLookup

[KEYBOARD]
KEY_UP = 105 0 0
KEY_PAUSE = 1073741906 0 0
END

Up arrow now initiates pause.

So we need a new keycode list
aka https://wiki.libsdl.org/SDLKeycodeLookup

**Comment by ghost** [comment]
well i knew the macro was doing something :) another mystery solved welcome to the sdl2 update
